api-documentation/shacl.md

```diff
@@ -0,0 +1,181 @@
+# Using SHACL to describe data structures
+
+## Story
+
+As an API publisher
+
+I want to use SHACL to describe data structures
+
+So that consumers can use existing tooling to consume the API
+
+## Details
+
+SHACL, which stands for Shapes Constraint Language is an RDF vocabulary, defined in a [W3C recommendation][shacl], which is used to describe graph shapes.
+
+It is very expressive, more rich than `hydra:Class` et.al., and out of the box provides multiple features to provide an accurate description of request payloads.
+
+[Many uses for SHACL have been proposed][ucr] such as, but not limited to, validation and building user interfaces dynamically.
+
+This page shows how Shapes can be used to replace or extend Hydra Core vocabulary terms within the API Documentation and resource representations.
+
+[shacl]: https://www.w3.org/TR/shacl/
+[ucr]: https://www.w3.org/TR/shacl-ucr/
+
+### Announce SHACL in API Documentation
+
+An API which chooses to use SHACL in addition to the built-in `hydra:Class` `MUST` announce that fact in its API Documentation resource
+
+```json
+{
+  "@context": {
+    "hydra": "http://www.w3.org/ns/hydra/core#",
+    "requires": { "@id": "hydra:requires", "@type": "@id" }
+  },
+  "@id": "/api",
+  "@type": "hydra:ApiDocumentation",
+  "requires": "http://www.w3.org/ns/hydra/shacl"
+}
+```
+
+By adding the `requires` property with the `hydra/shacl` identifier the client will be prepared to find SHACL terms within the payloads coming from this API.
+
+### Describing request bodies
+
+The most obvious use for Shapes is to use them to annotate supported operations' payloads. They will simply replace instances of `hydra:Class` used as objects of `hydra:expects`.
+
+```json
+{
+  "@context": "/api/context.jsonld",
+  "@id": "/api",
+  "@type": "hydra:ApiDocumentation",
+  "supportedClass": [{
+    "@id": "mov:Movies",
+    "supportedOperation": {
+      "@type": "schema:CreateAction",
+      "method": "POST",
+      "expects": ["mov:Movie", "mov:NewMovieShape"]
```

Comment:
Keeps `"mov:Movie"` to keep backwards compatibility

Reply:
This part is kind of tricky. What does the set means: is it a union or alternative?

Reply:
Union, obviously. Nothing tricky. Read it like:

> The operation expects `mov:Movie` which should conform to `mov:NewMovieShape`

Reply:
In this very case it may be obvious, but in general it's not. Imagine a case when an operation expects i.e. good or service (which are in many cases disjoined classes). Intent was to tell the client that the operation (basket?) expects a resource of one of those types - resource being both of the classes would violate the vocabulary)

Reply:
I don't see how this is relevant. Hydra Core would need to provide a way for alternatives just like OWL and SHACL do. Otherwise a set of objects is a union just like I would expect it anywhere else

Reply:
>Otherwise a set of objects is a union just like I would expect it anywhere else

It's not that straight forward. From pure RDF perspective it's a set of values and there is nothing supporting a statement a resource should be of all of the provided values. I'd actually prefer 'one of the possible values' rather than 'all of them'. I believe we need to specify it in the spec to avoid discrepancies.

Reply:
Please don't make me look up deep in some RDF specs but I'm pretty sure that the semantics would be precisely that. If you have multiple objects of a property, then they form a set.

This is how OWL works and also how SHACL works. Hence the provide `owl:oneOf` and `sh:or` respectively to explicitly describe alternative choices without ambiguity.

If that is not enough, let's look at `rdf:type`. `<Karol> rdf:type schema:Person, foaf:Person` means that that resource has both of those types, not ***one of the possible values***. Trying to apply different semantics to `expects` would IMO violate the principle of least astonishment.